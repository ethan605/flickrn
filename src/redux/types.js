export const PREFIXES = {
  FLICKR: 'flickr/',
  SHARED: 'shared/',
};

export const FLICKR = {
  CLEAR_PHOTO_INFO: 'flickr/CLEAR_PHOTO_INFO',
  GET_PHOTO_INFO: 'flickr/GET_PHOTO_INFO',
  SEARCH_PHOTOS: 'flickr/SEARCH_PHOTOS',
};

export const SHARED = {
};
