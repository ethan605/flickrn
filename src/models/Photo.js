import Joi from 'react-native-joi';

// Locals
import BaseModel, { COMMON_SCHEMAS } from './BaseModel';

const { INTEGER, STRING } = COMMON_SCHEMAS;

export default class Photo extends BaseModel {
  static dataValidator(): Joi.SchemaMap {
    return {
      farm: INTEGER.required(),
      id: STRING.required(),
      owner: STRING.required(),
      secret: STRING.required(),
      server: STRING.required(),
      title: STRING.required(),
    };
  }

  get thumbnail() {
    const { farm, id, secret, server } = this;
    return `https://farm${farm}.staticflickr.com/${server}/${id}_${secret}_b.jpg`;
  }
}
