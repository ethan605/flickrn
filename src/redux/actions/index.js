import * as flickr from './flickr';

/* eslint-disable import/prefer-default-export */

export { flickr };
