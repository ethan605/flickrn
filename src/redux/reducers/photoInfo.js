import _ from 'lodash';
import { FLICKR } from 'src/redux/types';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

// Models
import PhotoInfo from 'src/models/PhotoInfo';

const INITIAL_STATE = {
  data: undefined,
  errorMessage: undefined,
  loading: false,
};

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function photoInfoReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === FLICKR.CLEAR_PHOTO_INFO)
    return INITIAL_STATE;

  if (type === FLICKR.GET_PHOTO_INFO + ERROR) {
    const { error: { message: errorMessage } } = action;
    return { ...state, errorMessage, loading: false };
  }

  if (type === FLICKR.GET_PHOTO_INFO + SUCCESS) {
    const { payload: { photo } } = action;
    if (_.isEmpty(photo)) return state;

    return { ...state, data: PhotoInfo.build(photo), loading: false };
  }

  if (type === FLICKR.GET_PHOTO_INFO)
    return { ...state, data: undefined, errorMessage: undefined, loading: true };

  return state;
}
