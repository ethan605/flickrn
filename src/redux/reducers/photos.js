import _ from 'lodash';
import { FLICKR } from 'src/redux/types';

// Constants
import { AXIOS_REQUEST_SUFFIXES } from 'src/constants';

// Models
import Photo from 'src/models/Photo';

const INITIAL_STATE = {
  currentPage: 1,
  data: [],
  errorMessage: undefined,
  hasMore: true,
  loading: false,
};

const { ERROR, SUCCESS } = AXIOS_REQUEST_SUFFIXES;

export default function photosReducer(state = INITIAL_STATE, action) {
  const { type } = action;

  if (type === FLICKR.SEARCH_PHOTOS + ERROR) {
    const { error: { message: errorMessage } } = action;
    return { ...state, errorMessage, hasMore: false, loading: false };
  }

  if (type === FLICKR.SEARCH_PHOTOS + SUCCESS) {
    const { payload: { photos } } = action;
    if (_.isEmpty(photos)) return state;

    const { page, pages, photo } = photos;
    const hasMore = page < pages;

    const { data: currentPhotos } = state;
    const newPhotos = Photo.buildArray(photo);

    const data = page === 1 ? newPhotos : [...currentPhotos, ...newPhotos];
    return { ...state, hasMore, data, currentPage: page, loading: false };
  }

  if (type === FLICKR.SEARCH_PHOTOS)
    return { ...state, errorMessage: undefined, loading: true };

  return state;
}
