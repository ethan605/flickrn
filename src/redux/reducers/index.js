import photoInfo from './photoInfo';
import photos from './photos';

export { photoInfo, photos };
