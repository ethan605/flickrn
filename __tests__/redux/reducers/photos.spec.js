import { photos as photosReducer } from 'src/redux/reducers';
import { FLICKR } from 'src/redux/types';

// Helpers
import buildMockState from '__tests__/helpers/buildMockState';
import { buildAxiosActions } from '__tests__/helpers/mockApiRequests';

// Models
import Photo from 'src/models/Photo';

describe('redux/reducers/photos', () => {
  const { photos: mockState } = buildMockState();
  const mockSuccess = require('__tests__/fixtures/searchPhotos/success.json');
  const mockError = require('__tests__/fixtures/searchPhotos/error.json');

  const requestAction = { type: FLICKR.SEARCH_PHOTOS };
  
  const { photos } = mockSuccess;
  const { errorAction, successAction } = buildAxiosActions(requestAction, { photos }, mockError);

  test('start', () => {
    const currentState = {
      ...mockState,
      errorMessage: 'some random error',
    };

    const expectedState = {
      ...mockState,
      loading: true,
    };

    const reduced = photosReducer(currentState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('success', () => {
    const currentState = {
      ...mockState,
      loading: true,
    };

    const expectedState = {
      ...mockState,
      data: Photo.buildArray(mockSuccess.photos.photo),
    };

    const reduced = photosReducer(currentState, successAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('error', () => {
    const currentState = {
      ...mockState,
      loading: true,
    };

    const expectedState = {
      ...mockState,
      errorMessage: 'Random flickr.photos.search error',
      hasMore: false,
    };

    const reduced = photosReducer(currentState, errorAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });

  test('refresh', () => {
    const data = Photo.buildArray(mockSuccess.photos.photo);
    
    const currentState = { ...mockState, data };
    const expectedState = { ...mockState, data, loading: true };

    const reduced = photosReducer(currentState, requestAction);
    expect(reduced).toEqual(expectedState);
    expect(reduced).toMatchSnapshot();
  });
});
