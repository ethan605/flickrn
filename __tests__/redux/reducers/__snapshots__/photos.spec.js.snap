// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux/reducers/photos error 1`] = `
Object {
  "currentPage": 1,
  "data": Array [],
  "errorMessage": "Random flickr.photos.search error",
  "hasMore": false,
  "loading": false,
}
`;

exports[`redux/reducers/photos refresh 1`] = `
Object {
  "currentPage": 1,
  "data": Array [
    Photo {
      "farm": 5,
      "id": "40107995774",
      "owner": "152508136@N06",
      "secret": "7ed5905c16",
      "server": "4778",
      "title": "🐱🐶🐶 💖",
    },
    Photo {
      "farm": 5,
      "id": "26945278248",
      "owner": "128512949@N07",
      "secret": "9d9fc82c4c",
      "server": "4777",
      "title": "@Cgshapes , @Ella , @Fashionnatic , @Serendipity , @Armonia",
    },
    Photo {
      "farm": 5,
      "id": "40772774262",
      "owner": "52361229@N06",
      "secret": "24942a6a29",
      "server": "4771",
      "title": "These Guys starting 3 days early.....disclaimer: no pets or dogs were hurt or got intoxicated during the making of this image.",
    },
    Photo {
      "farm": 5,
      "id": "38999884280",
      "owner": "162254693@N02",
      "secret": "0e07ee8b6e",
      "server": "4784",
      "title": "Corgi king",
    },
    Photo {
      "farm": 5,
      "id": "39914583135",
      "owner": "113454068@N06",
      "secret": "25e007537b",
      "server": "4783",
      "title": "Sleeping Chorgi",
    },
  ],
  "errorMessage": undefined,
  "hasMore": true,
  "loading": true,
}
`;

exports[`redux/reducers/photos start 1`] = `
Object {
  "currentPage": 1,
  "data": Array [],
  "errorMessage": undefined,
  "hasMore": true,
  "loading": true,
}
`;

exports[`redux/reducers/photos success 1`] = `
Object {
  "currentPage": 1,
  "data": Array [
    Photo {
      "farm": 5,
      "id": "40107995774",
      "owner": "152508136@N06",
      "secret": "7ed5905c16",
      "server": "4778",
      "title": "🐱🐶🐶 💖",
    },
    Photo {
      "farm": 5,
      "id": "26945278248",
      "owner": "128512949@N07",
      "secret": "9d9fc82c4c",
      "server": "4777",
      "title": "@Cgshapes , @Ella , @Fashionnatic , @Serendipity , @Armonia",
    },
    Photo {
      "farm": 5,
      "id": "40772774262",
      "owner": "52361229@N06",
      "secret": "24942a6a29",
      "server": "4771",
      "title": "These Guys starting 3 days early.....disclaimer: no pets or dogs were hurt or got intoxicated during the making of this image.",
    },
    Photo {
      "farm": 5,
      "id": "38999884280",
      "owner": "162254693@N02",
      "secret": "0e07ee8b6e",
      "server": "4784",
      "title": "Corgi king",
    },
    Photo {
      "farm": 5,
      "id": "39914583135",
      "owner": "113454068@N06",
      "secret": "25e007537b",
      "server": "4783",
      "title": "Sleeping Chorgi",
    },
  ],
  "errorMessage": undefined,
  "hasMore": true,
  "loading": false,
}
`;
